ndk/src/midi.rs

```diff
@@ -0,0 +1,286 @@
+//! Bindings for [`AMidiDevice`], [`AMidiInputPort`], and [`AMidiOutputPort`]
+//!
+//! See [the NDK guide](https://developer.android.com/ndk/guides/audio/midi) for
+//! design and usage instructions, and [the NDK reference](https://developer.android.com/ndk/reference/group/midi)
+//! for an API overview.
+//!
+//! [`AMidiDevice`]: https://developer.android.com/ndk/reference/group/midi#amididevice
+//! [`AMidiInputPort`]: https://developer.android.com/ndk/reference/group/midi#amidiinputport
+//! [`AMidiOutputPort`]: https://developer.android.com/ndk/reference/group/midi#amidioutputport
+#![cfg(feature = "midi")]
+
+pub use super::media::Result;
+use super::media::{construct, NdkMediaError};
+
+use num_enum::{IntoPrimitive, TryFromPrimitive};
+use std::fmt;
+use std::marker::PhantomData;
+use std::os::raw::{c_int, c_uint};
+use std::ptr::NonNull;
+
+/// Result of [`MidiOutputPort::receive`].
+#[derive(Copy, Clone, Debug)]
+#[repr(u32)]
+pub enum MidiOpcode {
+    /// No MIDI messages are available.
+    NoMessage,
+    /// Received a MIDI message with the given length.
+    Data(usize),
+    /// Instructed to discard all pending MIDI data.
+    Flush,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, TryFromPrimitive, IntoPrimitive)]
+#[repr(u32)]
+pub enum MidiDeviceType {
+    Bluetooth = ffi::AMIDI_DEVICE_TYPE_BLUETOOTH,
+    USB = ffi::AMIDI_DEVICE_TYPE_USB,
+    Virtual = ffi::AMIDI_DEVICE_TYPE_VIRTUAL,
+}
+
+#[derive(Debug)]
+pub struct MidiDevice {
+    inner: NonNull<ffi::AMidiDevice>,
+}
+
+impl MidiDevice {
+    /// Creates an `MidiDevice` from a pointer.
+    ///
+    /// # Safety
+    /// By calling this function, you assert that the pointer is a valid pointer to a native
+    /// `AMidiDevice`.
+    pub unsafe fn from_ptr(inner: NonNull<ffi::AMidiDevice>) -> Self {
+        Self { inner }
+    }
+
+    fn as_ptr(&self) -> *mut ffi::AMidiDevice {
+        self.inner.as_ptr()
+    }
+
+    /// Connects a native Midi Device object to the associated Java MidiDevice object.
+    ///
+    /// Use this AMidiDevice to access the rest of the native MIDI API.
```

Comment:
```suggestion
    /// Use this [`MidiDevice`] to access the rest of the native MIDI API.
```

And even better: replace `this` with `the returned`.

Reply:
Fixed!